### PR TITLE
Plugins: Example CLI Plugin, a Resource Lister

### PR DESCRIPTION
**Pull request by clintoncwolfe**
Included in this PR is an example of Plugins v2 CliCommand plugin. It is very simple - it simply lists the names of the resources that InSpec knows about.

The plugin is exhaustively commented for pedagogic purposes.

The kit includes:
* an implementation of an InSpec CLI Command, using the InSpec PluginV2 API
* documentation in the README
* tests, at the unit and functional level
* a .gemspec, for packaging and publishing it as a gem
* a Gemfile, for managing its dependencies
* a Rakefile, for running development tasks
* Rubocop linting support for using the base InSpec project rubocop.yml (See Rakefile)

Additionally, there are two isolated commits that impact InSpec, outside of the example, both providing better support for an independent project to rely on the inspec installation for development support. [5188182](https://github.com/inspec/inspec/pull/3421/commits/51881823e57ac0ef2ed83b199e61e5cc80b15558) is straightforward, and [fff30db](https://github.com/inspec/inspec/pull/3421/commits/fff30db206ffd8c4cc9b8cca4f44ba9da499c4f9) is hair-raising.  Should either prove to be controversial, they can be cherry-picked into a separate PR.

I believe we should publish this example plugin to rubygems, after approval.

This PR supercedes #3400.

### CODE REVIEW

**Review by jquick — APPROVED** [review]
Thanks @clintoncwolfe. Lots of good info here. Very nice job explaining every part of the plugin.

File: .rubocop.yml
Comment (jerryaldrichiii) [inline .rubocop.yml:9]: I'm not following this comment. The wildcards below seem to have nothing to do with `examples/plugins/*/lib`

File: examples/plugins/inspec-resource-lister/README.md
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/README.md]: Should this be `you can do so`? Also, I would reword to `just use this` and or `(and not learn how to write a plugin)`, but I'll leave those to you.

File: examples/plugins/inspec-resource-lister/README.md
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/README.md:35]: Industrial grade...heh :smile:

File: examples/plugins/inspec-resource-lister/README.md
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/README.md]: `InSpec`?

File: examples/plugins/inspec-resource-lister/Rakefile
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/Rakefile:32]: I think it's `RuboCop`...but I'm not sure if we follow that internally.

File: examples/plugins/inspec-resource-lister/Rakefile
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/Rakefile:34]: `InSpec`?

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:22]: Capital `Plugin`?

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:27]: Hyphens confuse me...this could be correct though.

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:42]: Could we check for this in code?

File: examples/plugins/inspec-resource-lister/lib/inspec-resource-lister/plugin.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/lib/inspec-resource-lister/plugin.rb]: `InSpec`

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:26]: RuboCop would complain about double quotes without interpolation. The parent `describe` block is correct...but the rest seem to use double quotes. I can ignore it, but I'd like to see the same used within this file.

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:59]: Should this be `let(:outcome)`

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:77]: There are newlines between blocks elsewhere in the file. I'll leave it to you though.

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb]: Should this be `let(:outcome)`?

File: examples/plugins/inspec-resource-lister/test/helper.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb]: Should be `InSpec`

File: examples/plugins/inspec-resource-lister/test/helper.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb]: Should be `i.e.`

File: examples/plugins/inspec-resource-lister/test/helper.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb]: Should be `appearing`

File: examples/plugins/inspec-resource-lister/test/helper.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/helper.rb:25]: Hmmm...I disagree with this. I don't think we should commit requires for `pry/byebug`/

File: examples/plugins/inspec-resource-lister/test/unit/README.md
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/README.md]: Looks like you capitalize after the hyphen in the sentence above, could we here too? Also, should be `commands are`.

File: examples/plugins/inspec-resource-lister/test/unit/README.md
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/README.md]: Seems capitalization varies throughout the docs here. I was expecting `What are Unit Tests?`.

File: examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb]: Should be `command`? Or, maybe `Command`? Seems to vary in other docs.

File: examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb
Comment (jerryaldrichiii) [inline examples/plugins/inspec-resource-lister/test/unit/cli_args_test.rb:34]: Also mixed usage of quote styles in this file.

File: lib/plugins/shared/core_plugin_test_helper.rb
Comment (jerryaldrichiii) [inline lib/plugins/shared/core_plugin_test_helper.rb]: Muhammad Ali level fancy footwork going on here.

**Review by jerryaldrichiii — CHANGES_REQUESTED** [review]
Great work @clintoncwolfe.  There is a superb amount of documentation here.

As usual, I only have grammatical/style requests. I'm also a bit troubled by the line breaks in comments not being consistent...but I won't hold up the PR over it.

File: .rubocop.yml
Comment (jquick) [reply .rubocop.yml:9]: Right we are adding everything BUT `examples/plugins/*/lib` to the Exclude list here.

Comment (jerryaldrichiii) [reply .rubocop.yml:9]: :man_facepalming:

File: examples/plugins/inspec-resource-lister/README.md
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/README.md]: Refers to the identifier of the repo, not the name of the product.  I'll quote it.

File: examples/plugins/inspec-resource-lister/Rakefile
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/Rakefile:34]: No, this refers to a gem name, which is literally 'inspec'.

File: examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/inspec-resource-lister.gemspec:42]: This is a template / example.  We will have no code exposure to it until they try to run it against inspec.

If they used too low of a version (which we cannot go back in time and add checks for) it will fail because they are trying to use a plugin system that does not exist.

File: examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/test/functional/inspec_resource_lister_test.rb:26]: Please use the team's agreed-upon guidelines for where and how we apply style rules.

File: examples/plugins/inspec-resource-lister/test/helper.rb
Comment (clintoncwolfe) [reply examples/plugins/inspec-resource-lister/test/helper.rb:25]: Please bring up an argument for your position in the next standup.

**Review by jerryaldrichiii — APPROVED** [review]

**Pull request merged**
